### DISCUSSION TITLE
Connecting non SENEC systems / wallboxes

### DISCUSSION THREAD

**Discussion by loebse** [solectrus/solectrus#1821] [General]
First of all: thanks a lot @ledermann  for this great dashboard! I was able to host it on docker in minutes and it just looks fantastic!

### My hardware prerequisites

- SENEC.Home V3 hybrid duo (11.4 kWp)
- eon eBox Professional (connected vio MODBUS)
- Raspberry Pi (running ioBroker, Docker, etc.)

### The Challenge
As my wallbox is not recognized as wallbox in terms of SENEC, I can only display the total house consumption in the dashboard. There is no way to separate the wallbox consumption from the rest of the house consumption

### First attempt (fail)
My first attempt was to set the "wallbox_charging_power" in the InfluxDB  directly using the value of the wallbox that I have available via MODBUS in ioBroker. However, this was leading to the issue that the wallbox power was added to the house consumption which already contained the wallbox consumption --> fail

```mermaid
flowchart LR
    SENEC --> senec-collector --> InfluxDB --> Dashboard
    Wallbox --> ioBroker --> MQTT --> InfluxDB
```

### Second attempt
My second idea was to just not use the "senec-collector" at all but write the value of all fields into the Influxdb directly. To do so, I tried the following setup:

```mermaid
flowchart LR
    SENEC --> ioBroker 
    Wallbox --> ioBroker --> MQTT --> Telegraf --> InfluxDB --> Dashboard
```

- ioBroker on the Raspberry Pi is reading the SENEC data via LAN and the Wallbox data via MODBUS
- ioBroker calculates housepower - wallboxpower
- ioBroker publishes all values via MQTT
- Telegraf (with MQTT plugin) subscribes to the published topics and writes the values into the Influx DB

This solution looks pretty stable to me and does what I need.

### Third attempt
My third attempt (not finished yet) is to create a more generic program (probably in GO) to read the required values from MQTT and write it into the Influx db. This should be very flexible and would also allow to connect non-SENEC systems as long as you can publish the values via MQTT.

```mermaid
flowchart LR
    SENEC --> ioBroker 
    Wallbox --> ioBroker --> MQTT --> generic-MQTT-collector --> InfluxDB --> Dashboard
   Generic-System --> MQTT
```

As soon as I have some more time, I will work on the third attempt. Please feel free to provide some feddback :wink:

**Comment by ledermann** [comment]
This is great, @loebse! One of my favorite long-term goals is to support other photovoltaic systems or devices in addition to SENEC. I am extremely excited to see what you will create. Please feel free to use the programming language you are most comfortable with, such as Go. Ultimately, it will likely result in a Docker image regardless.

If you have any questions about the InfluxDB structure of SOLECTRUS, please let me know.

Just a small correction: You are talking about "senec-importer", but the correct name is "senec-collector". The [senec-importer](https://github.com/solectrus/senec-importer) allows importing CSV provided by `mein-senec.de`, while the [senec-collector](https://github.com/solectrus/senec-collector) accesses the SENEC device directly.

**Reply by loebse** [reply]
Thanks for the correction @ledermann , I changed it to "senec-collector"

**Comment by loebse** [comment]
I have created a first prototype that follows the third approach mentioned in the initial post. However, this still comes with a lot of hard coded values, some work arounds, and is missing a structured testing approach :see_no_evil:

@ledermann, I decided to create this in Ruby as your collector as well. Something that I have no experience with is the creation of the docker container. Can you maybe share some good resources on how to learn this?

**Reply by ledermann** [reply]
Wow, this sounds very promising! I'm excited to give this a try.

Regards Docker: I think you mean Docker Image, not Docker container, right? I use GitHub Actions for building and publishing Docker images, so this is part of the CI. See here: https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

My related code for the SENEC collector can be found here:
https://github.com/solectrus/senec-collector/blob/develop/.github/workflows/push.yml
It's a bit more extensive because QEMU is used to create Docker images for different platforms (ARM + AMD64).

Is your code available in a GitHub repo?

**Reply by loebse** [reply]
I was talking about an image and not a container, correct.

My code is on Github. It used to be a private repo but I changed it to public now: https://github.com/loebse/mqtt-collector

I did not manage to create a working Docker image as of now as there are a too many new things for me to learn with very limited time 😉 
I can successfully build the image but when starting it I just get the message `bundler: not executable: src/main.rb`

**Reply by ledermann** [reply]
Cool 💪 
I just sent a PR to fix some small issues: https://github.com/loebse/mqtt-collector/pull/2

Now I guess it's time for me to get busy with MQTT :-)

**Reply by loebse** [reply]
Awesome @ledermann! These changed made it for the docker image.

I will now do some changes to read all the relevant data via MQTT topics and apply the correct data structure so solectrus can consume the data.

**Comment by ledermann** [comment]
@loebse Some feedback from me:

Today, I set up [ioBroker](https://www.iobroker.net/), including the [SENEC adapter](https://github.com/nobl/ioBroker.senec) and activated the MQTT broker. Not a simple task, but finally the measurements from the SENEC storage are published via MQTT.

Next, I configured and started your MQTT Collector, which then fetched the data from ioBroker and wrote it into InfluxDB. There is just a minor issue: The field names for InfluxDB are incorrect, as you pointed out. A simple mapping is required.

Overall, this is truly fantastic :-)

**Reply by loebse** [reply]
Great to hear that it worked for you @ledermann 

I published an update in my repo today for the field mapping. Maybe not the most elegant way but at least I get the correct field names in the InfluxDB now!
